### ISSUE TITLE
Can prove Void using %reflection functions

### ISSUE THREAD

**Issue by david-christiansen** [idris-lang/Idris-dev#2716]
The semantics of the `%reflection` functions is a bit funky:
```
%reflection
ups : Nat -> Nat
ups (plus j k) = 42
ups (S k) = k
ups Z = 23

ups42 : (j, k : Nat) -> ups (plus j k) = 42
ups42 j k = Refl

inequal : the Nat 23 = the Nat 42 -> Void
inequal Refl impossible

broken : Void
broken = let x = ups42 Z Z in inequal x
```

**Comment by david-christiansen** [comment]
The problem here is that the functions are not, in fact, functions, because they can check for non-canonical forms.

**Comment by edwinb** [comment]
Probably %reflection functions should be classified differently by the totality checker (e.g. perhaps Reflection should be a result of totality checking along with Partial and Total) - they exist for a very specific purpose after all.

**Comment by david-christiansen** [comment]
I'm not sure that's enough... `%reflection` programs need to execute at compile time in order to be useful at all. Once they've executed, they should result in well-typed terms, but they themselves are allowed to do any abitrary thing (as I proved in the issue). It seems to me that we need some sort of stronger grip on staging - the `%reflection` programs must run first, and then the results are checked. This suggests to me that we might need some kind of `%runReflection` operator that causes them to reduce, and that doesn't type check at all itself, or something like that. Then they can be considered `Reflection` totality outside of that operator, at not run at all. `quoteGoal` could imply `%runReflection`.

**Comment by ahmadsalim** [comment]
@edwinb @david-christiansen  I guess this is less critical now, since FirstClassReflection is an extension.

**Comment by david-christiansen** [comment]
I agree that it's less critical if it relies on a language extension.

**Comment by ahmadsalim** [comment]
@david-christiansen I updated it to use the template. Do you think my updates are OK?
